src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs

```diff
@@ -2134,23 +2163,94 @@ void EmitAnchors(RegexNode node)
+                // want to unroll for shorter strings, because the overhead of invoking StartsWith instead of doing a few simple
+                // inline comparisons is very measurable, especially if we're doing a culture-sensitive comparison and StartsWith
+                // accesses CultureInfo.CurrentCulture on each call.  We need to be cognizant not only of the cost if the whole
+                // string matches, but also the cost when the comparison fails early on, and thus we pay for the call overhead
+                // but don't reap the benefits of all the vectorization StartsWith can do.
+                const int MaxUnrollLength = 64;
+                if (!caseInsensitive && // StartsWith(..., XxIgnoreCase) won't necessarily be the same as char-by-char comparison
+                    node.Str!.Length > MaxUnrollLength)
                 {
+                    // if (!textSpan.Slice(textSpanPos).StartsWith("...") goto doneLabel;
                     Ldloca(textSpanLocal);
-                    Ldc(textSpanPos + i);
-                    Call(s_spanGetItemMethod);
+                    Ldc(textSpanPos);
+                    Call(s_spanSliceIntMethod);
+                    Ldstr(node.Str);
+                    Call(s_stringAsSpanMethod);
+                    Call(s_spanStartsWith);
+                    BrfalseFar(doneLabel);
+                    textSpanPos += node.Str.Length;
+                    return;
+                }
+
+                // Emit the length check for the whole string.  If the generated code gets past this point,
+                // we know the span is at least textSpanPos + s.Length long.
+                ReadOnlySpan<char> s = node.Str;
+                EmitSpanLengthCheck(s.Length);
+
+                // If we're doing a case-insensitive comparison, we need to lower case each character,
+                // so we just go character-by-character.  But if we're not, we try to process multiple
+                // characters at a time; this is helpful not only for throughput but also in reducing
+                // the amount of IL and asm that results from this unrolling.
+                if (!caseInsensitive)
+                {
+                    // TODO https://github.com/dotnet/corefx/issues/39227:
+                    // If/when we implement CompileToAssembly, this code will either need to be special-cased
+                    // to not be used when saving out the assembly, or it'll need to be augmented to emit an
+                    // endianness check into the IL.  The code below is creating int/long constants based on
+                    // reading the comparison string at compile time, and the machine doing the compilation
+                    // could be of a different endianness than the machine running the compiled assembly.
+
+                    // On 64-bit, process 4 characters at a time until the string isn't at least 4 characters long.
+                    if (IntPtr.Size == 8)
```

Comment:
Why this check? Is it because you can be sure it won't cause a page fault (because the strings are aligned and pages are on 64 bit boundaries)? If so presumably the 2-by-2 block below can run even in 32 bit?

Reply:
Because Int64 loading/comparing in a 32-bit process can actually be a net loss.  We do similar checks in a bunch of places:
https://source.dot.net/#System.Private.CoreLib/IntPtr.cs,25b4f94fb6dd34fe,references

Reply:
I see. But the 2 at a time block could be included in a 32 bit process, right?

Reply:
Yes, and it is.

Reply:
Oh, I misread braces